ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryState.scala

```diff
@@ -17,8 +17,14 @@ private[memory] class InMemoryState private (log: MutableLog, state: MutableStat
   private val lockCurrentState = new StampedLock()
   @volatile private var lastLogEntryIndex = 0
 
-  def readLog[A](action: ImmutableLog => A): A =
-    action(log) // `log` is mutable, but the interface is immutable
+  def readLog[A](action: ImmutableLog => A): A = {
+    val stamp = blocking {
```

Comment:
Can we add a few lines as to why we need the lock?

Reply:
I think a comment on why we _didn't_ need the lock was necessary before. I am not convinced I need to explain why we need to lock.

Reply:
Added one anyway, because why not?

Reply:
My main confusion comes from the type that the action declares: `ImmutableLog => A`. Therefore I still don;t understand why do we need to use a lock for an immutable log.

Indeed our log is declared as `MutableLog`, but I assume some conversion is done when calling action which goes from mutable to immutable, therefore from what exact conditions do we protect ourselves against?

Reply:
`log` implements both `ImmutableLog` and `MutableLog`, and so can be cast to either. This means that without this lock, it could theoretically be read and written at the same time.

Reply:
> log implements both `ImmutableLog` and `MutableLog`
🤯

Reply:
This sort of pattern also often is meant to indicate to the reader that _the argument should only be used in the dynamic extent of the function call_; compare [typical `using` syntax](https://stackoverflow.com/questions/7443942/java-7-style-automatic-resource-management-for-scala).

Unlike `ST`, it doesn't have a type-level enforcement mechanism (consider just the consequences of passing `identity` as the argument), but that is how I would read any function that looks like continuation-passing in Scala.